### PR TITLE
build(release): configure semantic release

### LINKED ISSUE #288: Add a comment and a label on issues and PR released
To track easily the merged PR and closed issues without yet a release associated with them, it would be awesome to add a comment and a label on each issue and PR.
The label could only be something like released.
The comment could be more precise to know which release is associated to it.

I can provide a PR and was thinking to use [Semantic Release](https://semantic-release.gitbook.io/semantic-release/) FTW.

See original idea/need here https://github.com/actions/stale/issues/287#issuecomment-765253130

### PR DESCRIPTION
**Pull request by C0ZEN**
**Some infos:**
- [x] Install and configure a workflow to create a release on the master branch
	- Create a tag based on the commits
	- Generate the CHANGELOG.md
	- Set the tag in the package.json and package-lock.json
	- Create a git tag
	- Create a GitHub release
	- Add a comment and a label to each related issue and PR
- [x] Configure the semantic-pull-requests GitHub check
- [x] Configure the release GitHub workflow
	- When a push is done in master
		  - Get the current tag
		  - Create the release
		 	 - Install the dependencies (use some cache as well)
		 	 - Create the release (see above workflow)
		 	 - Create a new PR from master to main if there is a new tag (so that you can easily merge the changelog and packages changes)
 - [x] Configure the alpha release (alpha branch)

**What needs to be done before merging (by admin):**
- [ ] Install the [semantic-pull-requests](https://github.com/apps/semantic-pull-requests) check to disallow to merge a pull request with a bad commit syntax (required to avoid to fail the semantic release)
- [ ] Configure the [secrets in GitHub](https://github.com/semantic-release/github#github-authentication)

Closes #288

### CODE REVIEW

**Comment by C0ZEN** [comment]
@hross when you have the time if you can take a look 👍🏻.
Hard to test obviously but if you want to keep the stuff safe you can enable the dryRun option in the release config file to ensure that everything is ok.

**Comment by hross** [comment]
It's going to take me some time to get to this one, especially since it changes the release structure. I will take a look later this week and see if it makes sense (seems cool).

File: .github/workflows/release-alpha.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+            ${{ runner.os }}-build-${{ env.cache-name }}-
+            ${{ runner.os }}-build-
+            ${{ runner.os }}-
+      - name: Install dependencies
+        run: npm ci
+      - name: Release
+        run: npm run semver
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
```
Comment (C0ZEN) [inline .github/workflows/release-alpha.yml:34]: Not useful I think since there is no deployment on npm

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            ${{ runner.os }}-build-${{ env.cache-name }}-
+            ${{ runner.os }}-build-
+            ${{ runner.os }}-
+      - name: Install dependencies
+        run: npm ci
+      - name: Release
+        run: npm run semver
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
```
Comment (C0ZEN) [inline .github/workflows/release.yml:56]: Not useful I think since there is no deployment on npm

File: .releaserc.json [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        "linkReferences": true,
+        "commit": "commit",
+        "issue": "issues"
+      }
+    ],
+    [
+      "@semantic-release/changelog",
+      {
+        "changelogFile": "CHANGELOG.md",
+        "changelogTitle": "# Semantic Versioning Changelog"
```
Comment (C0ZEN) [inline .releaserc.json:36]: Maybe
```suggestion
        "changelogTitle": "# Stale Changelog"
```

File: .releaserc.json [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        "message": "build(release): new version ${nextRelease.version}"
+      }
+    ],
+    [
+      "@semantic-release/github",
+      {
+        "assets": [
+          {
+            "path": "dist/index.js",
+            "name": "stale-v${nextRelease.gitTag}.js",
```
Comment (C0ZEN) [inline .releaserc.json:56]:
```suggestion
            "name": "stale-${nextRelease.gitTag}.js",
```

File: .releaserc.json [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            "label": "Stale v${nextRelease.gitTag} release"
```
Comment (C0ZEN) [inline .releaserc.json:57]:
```suggestion
            "label": "Stale ${nextRelease.gitTag} release"
```
The v prefix will come with the one in tagFormat I think

File: .releaserc.json [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    ],
+    [
+      "@semantic-release/git",
+      {
+        "assets": [
+          "package.json",
+          "package-lock.json",
+          "CHANGELOG.md"
+        ],
+        "message": "build(release): new version ${nextRelease.version}"
```
Comment (C0ZEN) [inline .releaserc.json:47]:
```suggestion
        "message": "build(release): new version v${nextRelease.version}"
```
Not sure but I think the v prefix will not be included in the version, only in the tag

**Comment by hross** [comment]
I haven't gotten to this one yet. Will look next week.

**Comment by hross** [comment]
I still like this change but making release modifications requires me to spend a fair bit of time testing and watching and not breaking things so I have not yet merge this.

**Comment by hross** [comment]
I'm not sure if using angular commit message conventions is the way to go in order to manage the release process. Creating an alpha branch and auto-releasing would be pretty cool but keeping the release process manual means less commit maintenance and lock into the process, plus it's just a tag bump and more of a decision about a major release. Need to think more on this PR.

**Comment by C0ZEN** [comment]
@hross I am not sure to follow up with the issue.
The alpha branch would be optional in the process but to avoid automated release since you like to test the stuff internally it needs at least a new branch that is why I proposed the master one.
A release in that case is just a PR to merge either on alpha or on master (and is a good place to review the commits to make sure that it will create the tag you wish).
So IMO it's the same process as before except that the tag, the version, the GitHub release and the release notes would be generated based on the commits.
The drawback obviously is that there is a need to have good commits but since everything is squash on this project only the title of the PR matters.

**Comment by hross** [comment]
I don't think we should do this one for now. I don't want to commit this project to a different release strategy from our other actions right now. Maybe we can come back to it at some point, but for now I'm closing to avoid clutter.

**Pull request closed without merging**